In [1]:
import os
import sys
spark_path = os.environ['SPARK_HOME']
sys.path.append(spark_path + "/bin")
sys.path.append(spark_path + "/python")
sys.path.append(spark_path + "/python/pyspark/")
sys.path.append(spark_path + "/python/lib")
sys.path.append(spark_path + "/python/lib/pyspark.zip")
sys.path.append(spark_path + "/python/lib/py4j-0.10.9-src.zip")

import findspark
findspark.init()
import pyspark

In [2]:
number_cores = 6
memory_gb = 16
conf = (pyspark.SparkConf().setMaster('local[{}]'.format(number_cores)).set("spark.driver.maxResultSize", "5g"))
sc = pyspark.SparkContext(conf=conf)

In [3]:
!dir /users/trush/CSC496/Labs/Lab3

data  Lab3(1).ipynb  Lab3.ipynb


## How do you parse JSON?

In [4]:
import json

def parseJSON(t):
    res = json.loads(t)
    return res

test = '{"user_id":"ntlvfPzc8eglqvk92iDIAw","name":"Rafael","review_count":553,"yelping_since":"2007-07-06 03:27:11","useful":628,"funny":225,"cool":227,"elite":"","friends":"oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg, IkRib6Xs91PPW7pon7VVig, A8Aq8f0-XvLBcyMk2GJdJQ, eEZM1kogR7eL4GOBZyPvBA, e1o1LN7ez5ckCpQeAab4iw, _HrJVzFaRFUhPva8cwBjpQ, pZeGZGzX-ROT_D5lam5uNg, 0S6EI51ej5J7dgYz3-O0lA, woDt8raW-AorxQM_tIE2eA, hWUnSE5gKXNe7bDc8uAG9A, c_3LDSO2RHwZ94_Q6j_O7w, -uv1wDiaplY6eXXS0VwQiA, QFjqxXn3acDC7hckFGUKMg, ErOqapICmHPTN8YobZIcfQ, mJLRvqLOKhqEdkgt9iEaCQ, VKX7jlScJSA-ja5hYRw12Q, ijIC9w5PRcj3dWVlanjZeg, CIZGlEw-Bp0rmkP8M6yQ9Q, OC6fT5WZ8EU7tEVJ3bzPBQ, UZSDGTDpycDzrlfUlyw2dQ, deL6e_z9xqZTIODKqnvRXQ, 5mG2ENw2PylIWElqHSMGqg, Uh5Kug2fvDd51RYmsNZkGg, 4dI4uoShugD9z84fYupelQ, EQpFHqGT9Tk6YSwORTtwpg, o4EGL2-ICGmRJzJ3GxB-vw, s8gK7sdVzJcYKcPv2dkZXw, vOYVZgb_GVe-kdtjQwSUHw, wBbjgHsrKr7BsPBrQwJf2w, p59u2EC_qcmCmLeX1jCi5Q, VSAZI1eHDrOPRWMK4Q2DIQ, efMfeI_dkhpeGykaRJqxfQ, x6qYcQ8_i0mMDzSLsFCbZg, K_zSmtNGw1fu-vmxyTVfCQ, 5IM6YPQCK-NABkXmHhlRGQ, U_w8ZMD26vnkeeS1sD7s4Q, AbfS_oXF8H6HJb5jFqhrLw, hbcjX4_D4KIfonNnwrH-cg, UKf66_MPz0zHCP70mF6p1g, hK2gYbxZRTqcqlSiQQcrtQ, 2Q45w_Twx_T9dXqlE16xtQ, BwRn8qcKSeA77HLaOTbfiQ, jouOn4VS_DtFPtMR2w8VDA, ESteyJabbfvqas6CEDs3pQ","fans":14,"average_stars":3.57,"compliment_hot":3,"compliment_more":2,"compliment_profile":1,"compliment_cute":0,"compliment_list":1,"compliment_note":11,"compliment_plain":15,"compliment_cool":22,"compliment_funny":22,"compliment_writer":10,"compliment_photos":0}'
parseJSON(test)

{'user_id': 'ntlvfPzc8eglqvk92iDIAw',
 'name': 'Rafael',
 'review_count': 553,
 'yelping_since': '2007-07-06 03:27:11',
 'useful': 628,
 'funny': 225,
 'cool': 227,
 'elite': '',
 'friends': 'oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg, IkRib6Xs91PPW7pon7VVig, A8Aq8f0-XvLBcyMk2GJdJQ, eEZM1kogR7eL4GOBZyPvBA, e1o1LN7ez5ckCpQeAab4iw, _HrJVzFaRFUhPva8cwBjpQ, pZeGZGzX-ROT_D5lam5uNg, 0S6EI51ej5J7dgYz3-O0lA, woDt8raW-AorxQM_tIE2eA, hWUnSE5gKXNe7bDc8uAG9A, c_3LDSO2RHwZ94_Q6j_O7w, -uv1wDiaplY6eXXS0VwQiA, QFjqxXn3acDC7hckFGUKMg, ErOqapICmHPTN8YobZIcfQ, mJLRvqLOKhqEdkgt9iEaCQ, VKX7jlScJSA-ja5hYRw12Q, ijIC9w5PRcj3dWVlanjZeg, CIZGlEw-Bp0rmkP8M6yQ9Q, OC6fT5WZ8EU7tEVJ3bzPBQ, UZSDGTDpycDzrlfUlyw2dQ, deL6e_z9xqZTIODKqnvRXQ, 5mG2ENw2PylIWElqHSMGqg, Uh5Kug2fvDd51RYmsNZkGg, 4dI4uoShugD9z84fYupelQ, EQpFHqGT9Tk6YSwORTtwpg, o4EGL2-ICGmRJzJ3GxB-vw, s8gK7sdVzJcYKcPv2dkZXw, vOYVZgb_GVe-kdtjQwSUHw, wBbjgHsrKr7BsPBrQwJf2w, p59u2EC_qcmCmLeX1jCi5Q, VSAZI1eHDrOPRWMK4Q2DIQ, efMfeI_dkhpeGykaRJqxfQ, x6qYcQ8_i0mMDzSL

In [5]:
graph_data = sc.textFile("./data/yelp_academic_dataset_user.json")
graph_data.take(1)

['{"user_id":"ntlvfPzc8eglqvk92iDIAw","name":"Rafael","review_count":553,"yelping_since":"2007-07-06 03:27:11","useful":628,"funny":225,"cool":227,"elite":"","friends":"oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg, IkRib6Xs91PPW7pon7VVig, A8Aq8f0-XvLBcyMk2GJdJQ, eEZM1kogR7eL4GOBZyPvBA, e1o1LN7ez5ckCpQeAab4iw, _HrJVzFaRFUhPva8cwBjpQ, pZeGZGzX-ROT_D5lam5uNg, 0S6EI51ej5J7dgYz3-O0lA, woDt8raW-AorxQM_tIE2eA, hWUnSE5gKXNe7bDc8uAG9A, c_3LDSO2RHwZ94_Q6j_O7w, -uv1wDiaplY6eXXS0VwQiA, QFjqxXn3acDC7hckFGUKMg, ErOqapICmHPTN8YobZIcfQ, mJLRvqLOKhqEdkgt9iEaCQ, VKX7jlScJSA-ja5hYRw12Q, ijIC9w5PRcj3dWVlanjZeg, CIZGlEw-Bp0rmkP8M6yQ9Q, OC6fT5WZ8EU7tEVJ3bzPBQ, UZSDGTDpycDzrlfUlyw2dQ, deL6e_z9xqZTIODKqnvRXQ, 5mG2ENw2PylIWElqHSMGqg, Uh5Kug2fvDd51RYmsNZkGg, 4dI4uoShugD9z84fYupelQ, EQpFHqGT9Tk6YSwORTtwpg, o4EGL2-ICGmRJzJ3GxB-vw, s8gK7sdVzJcYKcPv2dkZXw, vOYVZgb_GVe-kdtjQwSUHw, wBbjgHsrKr7BsPBrQwJf2w, p59u2EC_qcmCmLeX1jCi5Q, VSAZI1eHDrOPRWMK4Q2DIQ, efMfeI_dkhpeGykaRJqxfQ, x6qYcQ8_i0mMDzSLsFCbZg, K_zSmtNGw1fu-vm

## Key points:
Always try and make your data smaller. Cache often if possible.

### Hands-on:
1. How do you apply pagerank on num_num?
2. How do you link the answers back to the actual page of hollies.dat?

In [6]:
from pyspark.sql import SQLContext
sqlContext = SQLContext(sc)

In [7]:
json_df = sqlContext.read.json("./data/yelp_academic_dataset_user.json")

In [8]:
%%time
json_df.count()

CPU times: user 3.82 ms, sys: 0 ns, total: 3.82 ms
Wall time: 3.72 s


1968703

In [9]:
%%time
json_df.count()

CPU times: user 2.61 ms, sys: 3.28 ms, total: 5.89 ms
Wall time: 2.7 s


1968703

In [10]:
json_df.printSchema()

root
 |-- average_stars: double (nullable = true)
 |-- compliment_cool: long (nullable = true)
 |-- compliment_cute: long (nullable = true)
 |-- compliment_funny: long (nullable = true)
 |-- compliment_hot: long (nullable = true)
 |-- compliment_list: long (nullable = true)
 |-- compliment_more: long (nullable = true)
 |-- compliment_note: long (nullable = true)
 |-- compliment_photos: long (nullable = true)
 |-- compliment_plain: long (nullable = true)
 |-- compliment_profile: long (nullable = true)
 |-- compliment_writer: long (nullable = true)
 |-- cool: long (nullable = true)
 |-- elite: string (nullable = true)
 |-- fans: long (nullable = true)
 |-- friends: string (nullable = true)
 |-- funny: long (nullable = true)
 |-- name: string (nullable = true)
 |-- review_count: long (nullable = true)
 |-- useful: long (nullable = true)
 |-- user_id: string (nullable = true)
 |-- yelping_since: string (nullable = true)



In [11]:
json_df.registerTempTable("tbl_json")

## 1. How many unique users are there in the dataset?

In [12]:
def parseJSONUser(x):
    res = json.loads(x)
    return res['user_id']

In [13]:
# map with just user_id .distinct().count()
users = graph_data.map(lambda x: parseJSONUser(x))
#users.take(1)
users.distinct().count()
#graph_data.countByKey()

1968703

There are 1,968,703 unique users in the dataset.

## 2. Identify the user(s) with the highest number of reviews. What is(are) the average/median stars for these user(s)?

In [14]:
def parseJSONReviews(x):
    res = json.loads(x)
    return (res['user_id'], res['name'], res['average_stars'], int(res['review_count']))

In [15]:
# map with user_id and review_count
user_data = graph_data.map(lambda x: parseJSON(x))
#user_data.take(1)
highest_num_reviews = graph_data.map(lambda x: parseJSONReviews(x)).sortBy(lambda x: (x[3]), ascending = False)
top10 = highest_num_reviews.take(10)
top10

[('8k3aO-mPeyhbR5HUucA5aA', 'Victor', 3.28, 14455),
 ('RtGqdDBvvBCjcu5dUqwfzA', 'Shila', 3.87, 12772),
 ('hWDybu_KvYLSdEFzGrniTw', 'Bruce', 3.64, 12487),
 ('Hi10sGSZNxQH3NLyWSZ1oA', 'Fox', 3.8, 11112),
 ('P5bUL3Engv-2z6kKohB6qQ', 'Kim', 3.8, 9875),
 ('8RcEwGrFIgkt9WQ35E6SnQ', 'George', 3.49, 7745),
 ('nmdkHL2JKFx55T3nq5VziA', 'Nijole', 3.71, 7626),
 ('Xwnf20FKuikiHcSpcEbpKQ', 'Kenneth', 3.32, 6762),
 ('CxDOIDnH8gp9KXzpBHJYXw', 'Jennifer', 3.33, 6633),
 ('HFECrzYDpgbS5EmTBtj2zQ', 'Eric', 3.93, 5500)]

In [16]:
# Average stars for users with top 10 highest number of reviews
def avg_top10(x):
    total = 0
    for i in range(10):
        total = total + top10[i][2]
    return total/10
# Median stars for users with top 10 highest number of reviews
def med_top10(x):
    x.sort(key = lambda x: float(x[2]))
    return x[len(x)//2][2]

In [17]:
print("Average: " + str(avg_top10(top10)))
print("Median: " + str(med_top10(top10)))

Average: 3.617
Median: 3.71


The top 10 users with the highest number of reviews are:
1. 8k3aO-mPeyhbR5HUucA5aA (Victor) with 3.28 average stars and 14455 reviews
2. RtGqdDBvvBCjcu5dUqwfzA (Shila) with 3.87 average stars and 12772 reviews
3. hWDybu_KvYLSdEFzGrniTw (Bruce) with 3.64 average stars and 12487 reviews
4. Hi10sGSZNxQH3NLyWSZ1oA (Fox) with 3.8 average stars and 11112 reviews
5. P5bUL3Engv-2z6kKohB6qQ (Kim) with 3.8 average stars and 9875 reviews
6. 8RcEwGrFIgkt9WQ35E6SnQ (George) with 3.49 average stars and 7745 reviews
7. nmdkHL2JKFx55T3nq5VziA (Nijole) with 3.71 average stars and 7626 reviews
8. Xwnf20FKuikiHcSpcEbpKQ (Kenneth) with 3.32 average stars and 6762 reviews
9. CxDOIDnH8gp9KXzpBHJYXw (Jennifer) with 3.33 average stars and 6633 reviews
10. HFECrzYDpgbS5EmTBtj2zQ (Eric) with 3.93 average stars and 5500 reviews

The average stars for the top 10 users is about 3.62 stars. The median stars for the top 10 users is 3.71 stars.

## 3. Identify several (two or more) interesting user attributes and think of and carry out one interesting analysis. Interesting is relative, and you should use Markdown cell to describe why/how these are interesting!

In [18]:
# Is the number of friends relative to number of fans correlated to elite status? Which is more important?
def parseJSON1(x):
    res = json.loads(x)
    return (res['user_id'], len(res['elite']), res['fans']/len(res['friends']))

# Do users with more elite years have higher or lower average stars?
def parseJSON2(x):
    res = json.loads(x)
    return (res['user_id'], len(res['elite']), res['average_stars'])

In [19]:
fans_friends_top = graph_data.map(lambda x: parseJSON1(x)).sortBy(lambda x: (x[2]), ascending = False)
fans_friends_top.take(10)

[('_nsd9q3GDLIu1vnBSg8yHQ', 0, 59.75),
 ('NhMvLlxJO0YNmi2-thbbzw', 0, 53.0),
 ('az7tv05nfECiV34AFr5cjw', 0, 51.0),
 ('qFCvUwkRBex6d3aAp86Z0g', 0, 50.5),
 ('5O7rAGUeVNqXjYxkXqSzqw', 0, 50.25),
 ('jcFd9jvJmZsQHQyTfuupVA', 0, 47.75),
 ('6aF1VL6ADlEWWl3erZZy9g', 0, 46.25),
 ('q5_VuD8-EZ8DlvLf8yEdCQ', 0, 41.25),
 ('idc5ItSIb262lqS-q89mJA', 0, 41.25),
 ('WoME7j9cEr-F6b3fk-0TcA', 0, 39.75)]

In [20]:
fans_friends_bottom = graph_data.map(lambda x: parseJSON1(x)).sortBy(lambda x: (x[2]), ascending = True)
fans_friends_bottom.take(10)

[('ttumcu6hWshk_EJVWrduDg', 0, 0.0),
 ('HwPGLzF_uXB3MF8bc5u5dg', 0, 0.0),
 ('cWw606W2PmcjTZrwLq8Hmg', 0, 0.0),
 ('Fta-vmTJYKGeVyyhm9D6Vw', 0, 0.0),
 ('B8z2bpAiT6IE5bq0o9FPEQ', 0, 0.0),
 ('61JaBatv-BJkzYol-ejEAg', 0, 0.0),
 ('Rm269GlzIZe8YINBeZ041g', 0, 0.0),
 ('rhQl_JFkCwrj0Rb8HUrNRQ', 0, 0.0),
 ('Gzn_5ZjbrOylXBkxuySY6Q', 0, 0.0),
 ('exeizgUKHDl926YLRS-5Mw', 0, 0.0)]

In [21]:
stars_top = graph_data.map(lambda x: parseJSON2(x)).sortBy(lambda x: (x[1]), ascending = False)
top20 = stars_top.take(20)

In [22]:
stars_bottom = graph_data.map(lambda x: parseJSON2(x)).sortBy(lambda x: (x[1]), ascending = True)
bot20 = stars_bottom.take(20)

In [23]:
sum1 = 0
sum2 = 0
for i in range(20):
    sum1 = sum1 + top20[i][2]
    sum2 = sum2 + bot20[i][2]
top_avg = sum1/20
bot_avg = sum2/20
print(top_avg)
print(bot_avg)

3.726
3.501


This analysis revolved around three questions:
1. Is the number of fans relative to the number of friends correlated to elite status?
3. Do users with more elite years have higher or lower average stars?

The answer to the first question lies in the number of elite years relative to the ratio of fans to friends. The question is essentially "Do elite squad members have more fans or friends and is this ratio relevant to elite status?" The data shows that there is generally no correlation between the ratio of fans to friends and elite status. Users with the highest ratio had 0 years as elite members and users with the lowest ratio also had zero years as elite members. This is interesting because it may initially be assumed that elite members may have more fans than friends, but there is no difference between elite and non-elite users in this metric.

The answer to the third question can be found by finding the top 20 (arbitrary number) users with the most elite years and comparing their average stars to the bottom 20 users with the least elite years. The average stars for the top 20 is 3.73 and 3.5 for the bottom 20. This indicates that Elite Squad member have a slightly higher average stars metric than non-Elite Squad members. Although there is just a slight difference in the average stars of elite and non-elite users, the results are interesting because they suggest that there may be a correlation in higher average reviews and elite status or that elite members are seeking out relatively better places to review than non-elite members.

## 4. Identify the top-ten most influential users from the Yelp dataset. Justify the attribute(s) that you pick to measure the level of influence!  

In [24]:
# PageRank: add weight for elite attribute
# Use numFriends and/or numFans according and/or average_stars to findings from 3

In [25]:
def parseJSONRanks(x):
    res = json.loads(x)
    return (res['user_id'], res['friends'].split(", "))

In [26]:
links = graph_data.map(lambda x: (parseJSONRanks(x)))
links.take(2)

[('ntlvfPzc8eglqvk92iDIAw',
  ['oeMvJh94PiGQnx_6GlndPQ',
   'wm1z1PaJKvHgSDRKfwhfDg',
   'IkRib6Xs91PPW7pon7VVig',
   'A8Aq8f0-XvLBcyMk2GJdJQ',
   'eEZM1kogR7eL4GOBZyPvBA',
   'e1o1LN7ez5ckCpQeAab4iw',
   '_HrJVzFaRFUhPva8cwBjpQ',
   'pZeGZGzX-ROT_D5lam5uNg',
   '0S6EI51ej5J7dgYz3-O0lA',
   'woDt8raW-AorxQM_tIE2eA',
   'hWUnSE5gKXNe7bDc8uAG9A',
   'c_3LDSO2RHwZ94_Q6j_O7w',
   '-uv1wDiaplY6eXXS0VwQiA',
   'QFjqxXn3acDC7hckFGUKMg',
   'ErOqapICmHPTN8YobZIcfQ',
   'mJLRvqLOKhqEdkgt9iEaCQ',
   'VKX7jlScJSA-ja5hYRw12Q',
   'ijIC9w5PRcj3dWVlanjZeg',
   'CIZGlEw-Bp0rmkP8M6yQ9Q',
   'OC6fT5WZ8EU7tEVJ3bzPBQ',
   'UZSDGTDpycDzrlfUlyw2dQ',
   'deL6e_z9xqZTIODKqnvRXQ',
   '5mG2ENw2PylIWElqHSMGqg',
   'Uh5Kug2fvDd51RYmsNZkGg',
   '4dI4uoShugD9z84fYupelQ',
   'EQpFHqGT9Tk6YSwORTtwpg',
   'o4EGL2-ICGmRJzJ3GxB-vw',
   's8gK7sdVzJcYKcPv2dkZXw',
   'vOYVZgb_GVe-kdtjQwSUHw',
   'wBbjgHsrKr7BsPBrQwJf2w',
   'p59u2EC_qcmCmLeX1jCi5Q',
   'VSAZI1eHDrOPRWMK4Q2DIQ',
   'efMfeI_dkhpeGykaRJqxfQ',
   'x6qYcQ8_i0m

In [27]:
N = links.count()
ranks = links.map(lambda line: (line[0], 1/N))
ranks.take(2)
#links.count()

[('ntlvfPzc8eglqvk92iDIAw', 5.079486342023149e-07),
 ('FOBRPlBHa3WPHFB5qYDlVg', 5.079486342023149e-07)]

In [28]:
votes = ranks.join(links)
votes.take(2)

[('ra0_P5KKYwU-imAVu_NW6g',
  (5.079486342023149e-07,
   ['icO65LTPSeMnvxw37zhRnw',
    'vsWDm2BuQ7lZIP1WwDhvcQ',
    'FNP8h7mywPGKkErQSBJdqg',
    'C-7EFRxfSrP6dzEKSB83cA',
    'CO-VVOYSThT_zvzPaEAJ6Q',
    'MUmclqmJqbnp-0n4RSFSrQ',
    'hey24H3dEiuOd3zZnOv30Q',
    'ZbHkj6HH4udAwi6MMx2SjA',
    'KjXGaWhQJQCi-MQQXwP1Xg',
    'ZrQBq71rn_ORyWqoivH82A',
    'HJldROBM71_LLKNtnih8ig',
    'yuSQQAZnzDkWYb6eFkQYJQ',
    'DWBgNQz5IPtdYX1zD3qr0A',
    '3-t_uaPKVQQmOUAIu7C-PA',
    'fafItEhuIkR8I0tmC3_vdA',
    '0Xl_Zrj8xaIu-clmND-toA',
    'SRnOg-4Ya4g4dUQ1x0Tdkg',
    'PNx1x8x3wjTIjrjhaNiVAQ',
    '2g0_rfFGk1k7GNDwfqtIqg',
    'ZaN-DYpDmiRVwNR7XGHTxg',
    'uNFyVo4e_1IynCIQIhevnQ',
    'nWoyOBL3z-4GHf0UflSn4w',
    'SfRsElsYGxIwuCjSNfNMjA',
    'Rrmb6UQUR-U3an6wy2h9wA',
    'zkZZqVTUgDglkOW0uegZNg',
    'ZcBtRj2R5LaQRxBZa5xEiQ',
    'QiOIFgh2jMssLgv6cL_xYA',
    'o5STsEtfvD1Ig0J7Z-1uxA',
    '-RsxLDYda6PJHVaoQlmcEw',
    'ZFxU8-XX9R2Oe1UgtV4Fdw',
    'HwmeT0Rq9OI4ox04WlNdAg',
    'tM1hqSXnb2B

In [29]:
def calculateVotes(t):
    res = []
    for item in t[1][1]:
        count = len(t[1][1])
        res.append((item, (t[1][0]/count)))
    return res

In [30]:
votes = ranks.join(links) \
        .flatMap(calculateVotes)
#votes.collect()

In [31]:
ranks.collect()

[('ntlvfPzc8eglqvk92iDIAw', 5.079486342023149e-07),
 ('FOBRPlBHa3WPHFB5qYDlVg', 5.079486342023149e-07),
 ('zZUnPeh2hEp0WydbAZEOOg', 5.079486342023149e-07),
 ('QaELAmRcDc5TfJEylaaP8g', 5.079486342023149e-07),
 ('xvu8G900tezTzbbfqmTKvA', 5.079486342023149e-07),
 ('z5_82komKV3mI4ASGe2-FQ', 5.079486342023149e-07),
 ('ttumcu6hWshk_EJVWrduDg', 5.079486342023149e-07),
 ('f4_MRNHvN-yRn7EA8YWRxg', 5.079486342023149e-07),
 ('UYACF30806j2mfbB5vdmJA', 5.079486342023149e-07),
 ('QG13XBbgHWydzThRBGJtyw', 5.079486342023149e-07),
 ('f6YuZP6iennHFVlnFJOXLQ', 5.079486342023149e-07),
 ('I_6wY8_RsewziNnKhGZg4g', 5.079486342023149e-07),
 ('q-v8elVPvKz0KvK69QSj1Q', 5.079486342023149e-07),
 ('HwPGLzF_uXB3MF8bc5u5dg', 5.079486342023149e-07),
 ('y4UuVowA9i3zj2hHyRMfHw', 5.079486342023149e-07),
 ('1WBxJ2r3A2QYfRSEzgcmkQ', 5.079486342023149e-07),
 ('-TT5e-YQU9xLb1JAGCGkQw', 5.079486342023149e-07),
 ('6bbHSJ0PrgSxh7e5nigKMw', 5.079486342023149e-07),
 ('4VmuXuSRhv5UxYUy3tMpiQ', 5.079486342023149e-07),
 ('pVU2DdtBF

In [32]:
ranks = votes.reduceByKey(lambda x, y: x + y).sortBy(lambda x: x[1], ascending = False)
ranks.take(10)

[('None', 0.42406142521242063),
 ('GGTF7hnQi6D5W77_qiKlqg', 0.00016013803395873016),
 ('Wc5L6iuvSNF5WGBlqIO8nw', 0.0001453671303478104),
 ('wd3xoNaDLib8dhQ7BxUl6g', 0.00013765373970909246),
 ('u3ZPMVVEzneq8x856WksJQ', 0.00012624046233580594),
 ('WeVkkF5L39888IPPlRhNpg', 0.00011611892139558681),
 ('NrSURtBigpxbdfL4nGB-Jg', 0.00011205626494020588),
 ('mGHP3esQQqCEicHLRV-g5Q', 0.00010656701415615383),
 ('NfU0zDaTMEQ4-X9dbQWd9A', 0.00010356022235849807),
 ('O_GWZZfQx7qv-n-CN7hsIA', 9.85859157946356e-05)]

The top 10 most influential users ranked by number of friends are:
1. None with a rank of 0.42406142521242063),
2. GGTF7hnQi6D5W77_qiKlqg with a rank of 0.00016013803395873016
3. Wc5L6iuvSNF5WGBlqIO8nw with a rank of 0.0001453671303478104
4. wd3xoNaDLib8dhQ7BxUl6g with a rank of 0.00013765373970909246
5. u3ZPMVVEzneq8x856WksJQ with a rank of 0.00012624046233580594
6. WeVkkF5L39888IPPlRhNpg with a rank of 0.00011611892139558681
7. NrSURtBigpxbdfL4nGB-Jg with a rank of 0.00011205626494020588
8. mGHP3esQQqCEicHLRV-g5Q with a rank of 0.00010656701415615383
9. NfU0zDaTMEQ4-X9dbQWd9A with a rank of 0.00010356022235849807
10. O_GWZZfQx7qv-n-CN7hsIA with a rank of 9.85859157946356e-05

This ranking is based off number of friends. Users with more friends have a wider sphere of influence, thus higher votes; therefore, users with many friends that have many friends are ranked higher.